Nullpointerexception in spring-integration-mqtt adapter while reconnecting

**In what version(s) of Spring Integration are you seeing this issue?**
5.5.3

**Describe the bug**

```
15:41:23.620 scheduler2 ERROR .s.s.s.TaskUtils$LoggingErrorHandler| Unexpected error occurred in scheduled task requestId=LnaKMnE6
java.lang.NullPointerException: null
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.connectAndSubscribe(MqttPahoMessageDrivenChannelAdapter.java:314)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.lambda$scheduleReconnect$2(MqttPahoMessageDrivenChannelAdapter.java:364)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
```	

A clear and concise description of what the bug is.

This line is throwing `NullPointerException` inside 
```
this.client.disconnectForcibly(this.disconnectCompletionTimeout);
```

**To Reproduce**
Write a simple unit test that makes adapter fails to connect to a topic.. This is the order of the exceptions

```
15:57:29.009 Thread-31 ERROR .MqttPahoMessageDrivenChannelAdapter| Error connecting or subscribing to [publish/*/*] 
org.eclipse.paho.client.mqttv3.MqttException: MqttException
	at org.eclipse.paho.client.mqttv3.internal.ExceptionHelper.createMqttException(ExceptionHelper.java:38)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms$ConnectBG.run(ClientComms.java:738)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.io.IOException: WebSocket Response header: Incorrect connection header
	at org.eclipse.paho.client.mqttv3.internal.websocket.WebSocketHandshake.receiveHandshakeResponse(WebSocketHandshake.java:157)
	at org.eclipse.paho.client.mqttv3.internal.websocket.WebSocketHandshake.execute(WebSocketHandshake.java:80)
	at org.eclipse.paho.client.mqttv3.internal.websocket.WebSocketSecureNetworkModule.start(WebSocketSecureNetworkModule.java:65)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms$ConnectBG.run(ClientComms.java:724)
	... 1 common frames omitted
```

then

```
15:57:59.016 Thread-31 ERROR .MqttPahoMessageDrivenChannelAdapter| Exception while connecting and subscribing, retrying 
org.eclipse.paho.client.mqttv3.MqttException: MqttException
	at org.eclipse.paho.client.mqttv3.internal.ExceptionHelper.createMqttException(ExceptionHelper.java:38)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms$ConnectBG.run(ClientComms.java:738)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.io.IOException: WebSocket Response header: Incorrect connection header
	at org.eclipse.paho.client.mqttv3.internal.websocket.WebSocketHandshake.receiveHandshakeResponse(WebSocketHandshake.java:157)
	at org.eclipse.paho.client.mqttv3.internal.websocket.WebSocketHandshake.execute(WebSocketHandshake.java:80)
	at org.eclipse.paho.client.mqttv3.internal.websocket.WebSocketSecureNetworkModule.start(WebSocketSecureNetworkModule.java:65)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms$ConnectBG.run(ClientComms.java:724)
	... 1 common frames omitted
```

```
15:58:09.079 scheduler1 ERROR .MqttPahoMessageDrivenChannelAdapter| Exception while unsubscribing 
org.eclipse.paho.client.mqttv3.MqttException: Client is not connected
	at org.eclipse.paho.client.mqttv3.internal.ExceptionHelper.createMqttException(ExceptionHelper.java:31)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms.sendNoWait(ClientComms.java:205)
	at org.eclipse.paho.client.mqttv3.MqttAsyncClient.unsubscribe(MqttAsyncClient.java:1244)
	at org.eclipse.paho.client.mqttv3.MqttClient.unsubscribe(MqttClient.java:551)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.doStop(MqttPahoMessageDrivenChannelAdapter.java:223)
	at org.springframework.integration.en
```

then

```
15:58:39.083 scheduler1 ERROR .MqttPahoMessageDrivenChannelAdapter| Error connecting or subscribing to [publish/*/*] 
org.eclipse.paho.client.mqttv3.MqttException: MqttException
	at org.eclipse.paho.client.mqttv3.internal.ExceptionHelper.createMqttException(ExceptionHelper.java:38)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms$ConnectBG.run(ClientComms.java:738)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.io.IOException: WebSocket Response header: Incorrect connection header
	at org.eclipse.paho.client.mqttv3.internal.websocket.WebSocketHandshake.receiveHandshakeResponse(WebSocketHandshake.java:157)
	at org.eclipse.paho.client.mqttv3.internal.websocket.WebSocketHandshake.execute(WebSocketHandshake.java:80)
	at org.eclipse.paho.client.mqttv3.internal.websocket.WebSocketSecureNetworkModule.start(WebSocketSecureNetworkModule.java:65)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms$ConnectBG.run(ClientComms.java:724)
	... 1 common frames omitted
```

then the NPE

```
java.lang.NullPointerException: null
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.connectAndSubscribe(MqttPahoMessageDrivenChannelAdapter.java:314)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.lambda$scheduleReconnect$2(MqttPahoMessageDrivenChannelAdapter.java:364)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
```	

				
	

Steps to reproduce the behavior.

**Expected behavior**

Library should never throw NPE. Looks like code is setting client to null and during reconnect using a null client to reconnect

A clear and concise description of what you expected to happen.

**Sample**

A link to a GitHub repository with a [minimal, reproducible sample](https://stackoverflow.com/help/minimal-reproducible-example).

Reports that include a sample will take priority over reports that do not.
At times, we may require a sample, so it is good to try and include a sample up front.

OK. I was able to reproduce an issue with this unit-test (based on the framework test environment):

```
	@Test
	public void testNoNPEOnReconnectAndStopRaceCondition() throws Exception {
		final IMqttClient client = mock(IMqttClient.class);
		MqttPahoMessageDrivenChannelAdapter adapter = buildAdapterIn(client, null, ConsumerStopAction.UNSUBSCRIBE_NEVER);
		adapter.setRecoveryInterval(10);

		willThrow(MqttException.class)
				.given(client)
				.subscribe(any(),ArgumentMatchers.<int[]>any());

		ThreadPoolTaskScheduler taskScheduler = new ThreadPoolTaskScheduler();
		AtomicReference<Throwable> reconnectException = new AtomicReference<>();
		CountDownLatch reconnectExceptionLatch = new CountDownLatch(1);
		taskScheduler.setErrorHandler(ex -> {
			reconnectException.set(ex);
			reconnectExceptionLatch.countDown();
		});
		taskScheduler.initialize();
		adapter.setTaskScheduler(taskScheduler);

		adapter.setApplicationEventPublisher(event -> {
			if (event instanceof MqttConnectionFailedEvent) {
				adapter.destroy();
			}
		});
		adapter.start();

		assertThat(reconnectExceptionLatch.await(10, TimeUnit.SECONDS)).isTrue();
		assertThat(reconnectException.get())
				.isNotNull()
				.isInstanceOf(MqttException.class);

		taskScheduler.destroy();
	}
```

(I believe the `MqttPahoMessageDrivenChannelAdapter` instance could be any).

The main point of this test that I call (as you do) an `adapter.destroy()` *within* the same thread as `MqttConnectionFailedEvent` emitted from that `connectAndSubscribe()`, so we are good to go to the `doStop()` and `null` the `client` property and therefore it leads to NPE on the next line in that `connectAndSubscribe()` which is `this.client.disconnectForcibly(this.disconnectCompletionTimeout);`.
The test result without any code changes is like this:
```
2021-09-09 13:06:04,509 ERROR [main] [org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter] - Error connecting or subscribing to [foo]
org.eclipse.paho.client.mqttv3.MqttException: MqttException
2021-09-09 13:06:04,557 ERROR [main] [org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter] - Exception while connecting and subscribing, retrying
org.eclipse.paho.client.mqttv3.MqttException: MqttException
2021-09-09 13:06:04,583 ERROR [ThreadPoolTaskScheduler-1] [org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter] - Error connecting or subscribing to [foo]
org.eclipse.paho.client.mqttv3.MqttException: MqttException

java.lang.AssertionError: 
Expecting actual throwable to be an instance of:
  org.eclipse.paho.client.mqttv3.MqttException
but was:
  java.lang.NullPointerException
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.connectAndSubscribe(MqttPahoMessageDrivenChannelAdapter.java:314)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.lambda$scheduleReconnect$2(MqttPahoMessageDrivenChannelAdapter.java:364)
```

Same NPE.
It is just because we are free to call any `synchronized` from the `synchronized` if we are withing the same execution flow on the same thread so far.

So, thank you for the report and stay tuned for the fix!

--- a/spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java
+++ b/spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java
@@ -84,6 +84,8 @@ public class MqttPahoMessageDrivenChannelAdapter extends AbstractMqttMessageDriv
 
 	private boolean manualAcks;
 
+	private ApplicationEventPublisher applicationEventPublisher;
+
 	private volatile IMqttClient client;
 
 	private volatile ScheduledFuture<?> reconnectFuture;
@@ -93,8 +95,6 @@ public class MqttPahoMessageDrivenChannelAdapter extends AbstractMqttMessageDriv
 	private volatile boolean cleanSession;
 
 	private volatile ConsumerStopAction consumerStopAction;
-
-	private ApplicationEventPublisher applicationEventPublisher;
 
 	/**
 	 * Use this constructor for a single url (although it may be overridden if the server
@@ -311,15 +311,17 @@ public class MqttPahoMessageDrivenChannelAdapter extends AbstractMqttMessageDriv
 				this.applicationEventPublisher.publishEvent(new MqttConnectionFailedEvent(this, ex));
 			}
 			logger.error(ex, () -> "Error connecting or subscribing to " + Arrays.toString(topics));
-			this.client.disconnectForcibly(this.disconnectCompletionTimeout);
-			try {
-				this.client.setCallback(null);
-				this.client.close();
+			if (this.client != null) { // Could be reset during event handling before
+				this.client.disconnectForcibly(this.disconnectCompletionTimeout);
+				try {
+					this.client.setCallback(null);
+					this.client.close();
+				}
+				catch (MqttException e1) {
+					// NOSONAR
+				}
+				this.client = null;
 			}
-			catch (MqttException e1) {
-				// NOSONAR
-			}
-			this.client = null;
 			throw ex;
 		}
 		finally {
@@ -355,25 +357,27 @@ public class MqttPahoMessageDrivenChannelAdapter extends AbstractMqttMessageDriv
 
 	private synchronized void scheduleReconnect() {
 		cancelReconnect();
-		try {
-			this.reconnectFuture = getTaskScheduler().schedule(() -> {
-				try {
-					logger.debug("Attempting reconnect");
-					synchronized (MqttPahoMessageDrivenChannelAdapter.this) {
-						if (!MqttPahoMessageDrivenChannelAdapter.this.connected) {
-							connectAndSubscribe();
-							MqttPahoMessageDrivenChannelAdapter.this.reconnectFuture = null;
+		if (isActive()) {
+			try {
+				this.reconnectFuture = getTaskScheduler().schedule(() -> {
+					try {
+						logger.debug("Attempting reconnect");
+						synchronized (MqttPahoMessageDrivenChannelAdapter.this) {
+							if (!MqttPahoMessageDrivenChannelAdapter.this.connected) {
+								connectAndSubscribe();
+								MqttPahoMessageDrivenChannelAdapter.this.reconnectFuture = null;
+							}
 						}
 					}
-				}
-				catch (MqttException ex) {
-					logger.error(ex, "Exception while connecting and subscribing");
-					scheduleReconnect();
-				}
-			}, new Date(System.currentTimeMillis() + this.recoveryInterval));
-		}
-		catch (Exception ex) {
-			logger.error(ex, "Failed to schedule reconnect");
+					catch (MqttException ex) {
+						logger.error(ex, "Exception while connecting and subscribing");
+						scheduleReconnect();
+					}
+				}, new Date(System.currentTimeMillis() + this.recoveryInterval));
+			}
+			catch (Exception ex) {
+				logger.error(ex, "Failed to schedule reconnect");
+			}
 		}
 	}
 
@@ -412,7 +416,7 @@ public class MqttPahoMessageDrivenChannelAdapter extends AbstractMqttMessageDriv
 				sendMessage(message);
 			}
 			catch (RuntimeException ex) {
-				logger.error(ex, () -> "Unhandled exception for " + message.toString());
+				logger.error(ex, () -> "Unhandled exception for " + message);
 				throw ex;
 			}
 		}

--- a/spring-integration-mqtt/src/test/java/org/springframework/integration/mqtt/MqttAdapterTests.java
+++ b/spring-integration-mqtt/src/test/java/org/springframework/integration/mqtt/MqttAdapterTests.java
@@ -22,6 +22,7 @@ import static org.assertj.core.api.Assertions.fail;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyLong;
 import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.ArgumentMatchers.isNull;
 import static org.mockito.BDDMockito.given;
 import static org.mockito.BDDMockito.willAnswer;
@@ -53,7 +54,6 @@ import org.aopalliance.intercept.MethodInterceptor;
 import org.assertj.core.api.Condition;
 import org.eclipse.paho.client.mqttv3.IMqttAsyncClient;
 import org.eclipse.paho.client.mqttv3.IMqttClient;
-import org.eclipse.paho.client.mqttv3.IMqttMessageListener;
 import org.eclipse.paho.client.mqttv3.IMqttToken;
 import org.eclipse.paho.client.mqttv3.MqttAsyncClient;
 import org.eclipse.paho.client.mqttv3.MqttCallback;
@@ -65,6 +65,7 @@ import org.eclipse.paho.client.mqttv3.MqttMessage;
 import org.eclipse.paho.client.mqttv3.MqttToken;
 import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;
 import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
 import org.mockito.ArgumentMatchers;
 import org.mockito.internal.stubbing.answers.CallsRealMethods;
 
@@ -541,8 +542,7 @@ public class MqttAdapterTests {
 		new DirectFieldAccessor(client).setPropertyValue("aClient", aClient);
 		willAnswer(new CallsRealMethods()).given(client).connect(any(MqttConnectOptions.class));
 		willAnswer(new CallsRealMethods()).given(client).subscribe(any(String[].class), any(int[].class));
-		willAnswer(new CallsRealMethods()).given(client).subscribe(any(String[].class), any(int[].class),
-				(IMqttMessageListener[]) isNull());
+		willAnswer(new CallsRealMethods()).given(client).subscribe(any(String[].class), any(int[].class), isNull());
 		willReturn(alwaysComplete).given(aClient).connect(any(MqttConnectOptions.class), any(), any());
 
 		IMqttToken token = mock(IMqttToken.class);
@@ -573,8 +573,51 @@ public class MqttAdapterTests {
 		verify(client).disconnectForcibly(5_000L);
 	}
 
+	@Test
+	public void testNoNPEOnReconnectAndStopRaceCondition() throws Exception {
+		final IMqttClient client = mock(IMqttClient.class);
+		MqttPahoMessageDrivenChannelAdapter adapter = buildAdapterIn(client, null, ConsumerStopAction.UNSUBSCRIBE_NEVER);
+		adapter.setRecoveryInterval(10);
+
+		MqttException mqttException = new MqttException(MqttException.REASON_CODE_SUBSCRIBE_FAILED);
+
+		willThrow(mqttException)
+				.given(client)
+				.subscribe(any(), ArgumentMatchers.<int[]>any());
+
+		LogAccessor logger = spy(TestUtils.getPropertyValue(adapter, "logger", LogAccessor.class));
+		new DirectFieldAccessor(adapter).setPropertyValue("logger", logger);
+		CountDownLatch exceptionLatch = new CountDownLatch(1);
+		ArgumentCaptor<MqttException> mqttExceptionArgumentCaptor = ArgumentCaptor.forClass(MqttException.class);
+		willAnswer(i -> {
+			exceptionLatch.countDown();
+			return null;
+		})
+				.given(logger)
+				.error(mqttExceptionArgumentCaptor.capture(), eq("Exception while connecting and subscribing"));
+
+		ThreadPoolTaskScheduler taskScheduler = new ThreadPoolTaskScheduler();
+		taskScheduler.initialize();
+		adapter.setTaskScheduler(taskScheduler);
+
+		adapter.setApplicationEventPublisher(event -> {
+			if (event instanceof MqttConnectionFailedEvent) {
+				adapter.destroy();
+			}
+		});
+		adapter.start();
+
+		assertThat(exceptionLatch.await(10, TimeUnit.SECONDS)).isTrue();
+		assertThat(mqttExceptionArgumentCaptor.getValue())
+				.isNotNull()
+				.isSameAs(mqttException);
+
+		taskScheduler.destroy();
+	}
+
 	private MqttPahoMessageDrivenChannelAdapter buildAdapterIn(final IMqttClient client, Boolean cleanSession,
-			ConsumerStopAction action) throws MqttException {
+			ConsumerStopAction action) {
+
 		DefaultMqttPahoClientFactory factory = new DefaultMqttPahoClientFactory() {
 
 			@Override
@@ -605,7 +648,7 @@ public class MqttAdapterTests {
 		DefaultMqttPahoClientFactory factory = new DefaultMqttPahoClientFactory() {
 
 			@Override
-			public IMqttAsyncClient getAsyncClientInstance(String uri, String clientId) throws MqttException {
+			public IMqttAsyncClient getAsyncClientInstance(String uri, String clientId) {
 				return client;
 			}
 

